File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/service/MapperService.kt
package com.hellotimepro.ktor.service

import com.hellotimepro.ktor.domain.Capsule
import com.hellotimepro.ktor.domain.User
import com.hellotimepro.ktor.dto.CapsuleDetail
import com.hellotimepro.ktor.dto.CapsuleListItem
import com.hellotimepro.ktor.dto.UserBrief
import com.hellotimepro.ktor.dto.UserDto
import java.time.OffsetDateTime
import java.time.ZoneOffset

/** 领域模型 → 响应 DTO。时间统一序列化为 ISO-8601 UTC（`...Z`）。 */
class MapperService {
    fun user(u: User): UserDto =
        UserDto(u.id.toString(), u.email, u.nickname, u.avatarId, iso(u.createdAt))

    fun detail(c: Capsule, owner: User, favoritedByMe: Boolean, includeContent: Boolean = true): CapsuleDetail {
        val opened = isOpened(c.openAt)
        return CapsuleDetail(
            id = c.id.toString(),
            code = c.code,
            title = c.title,
            creator = UserBrief(owner.nickname, owner.avatarId),
            openAt = iso(c.openAt),
            createdAt = iso(c.createdAt),
            inPlaza = c.inPlaza,
            favoriteCount = c.favoriteCount,
            isOpened = opened,
            content = if (opened && includeContent) c.content else null,
            favoritedByMe = favoritedByMe,
        )
    }

    fun listItem(c: Capsule, owner: User, favoritedByMe: Boolean, favoritedAt: OffsetDateTime?): CapsuleListItem {
        val opened = isOpened(c.openAt)
        val preview = if (opened) {
            val raw = c.content.trim()
            if (raw.length > 80) raw.substring(0, 80) + "…" else raw
        } else {
            null
        }
        return CapsuleListItem(
            id = c.id.toString(),
            code = c.code,
            title = c.title,
            creator = UserBrief(owner.nickname, owner.avatarId),
            openAt = iso(c.openAt),
            createdAt = iso(c.createdAt),
            inPlaza = c.inPlaza,
            favoriteCount = c.favoriteCount,
            isOpened = opened,
            favoritedByMe = favoritedByMe,
            favoritedAt = favoritedAt?.let { iso(it) },
            contentPreview = preview,
        )
    }

    private fun isOpened(openAt: OffsetDateTime): Boolean =
        !openAt.isAfter(OffsetDateTime.now(ZoneOffset.UTC))

    private fun iso(value: OffsetDateTime): String = value.toInstant().toString()
}

File: backends/aspnet/src/Services/MapperService.cs
using HelloTimePro.Aspnet.Domain;
using HelloTimePro.Aspnet.Dto;

namespace HelloTimePro.Aspnet.Services;

/// <summary>领域模型 → 响应 DTO。时间统一序列化为 ISO-8601 UTC（<c>...Z</c>）。单例、无状态。</summary>
public sealed class MapperService
{
    public UserDto User(User u) =>
        new(u.Id.ToString(), u.Email, u.Nickname, u.AvatarId, Times.Iso(u.CreatedAt));

    public CapsuleDetail Detail(Capsule c, User owner, bool favoritedByMe, bool includeContent = true)
    {
        var opened = IsOpened(c.OpenAt);
        return new CapsuleDetail(
            Id: c.Id.ToString(),
            Code: c.Code,
            Title: c.Title,
            Creator: new UserBrief(owner.Nickname, owner.AvatarId),
            OpenAt: Times.Iso(c.OpenAt),
            CreatedAt: Times.Iso(c.CreatedAt),
            InPlaza: c.InPlaza,
            FavoriteCount: c.FavoriteCount,
            IsOpened: opened,
            Content: opened && includeContent ? c.Content : null,
            FavoritedByMe: favoritedByMe);
    }

    public CapsuleListItem ListItem(Capsule c, User owner, bool favoritedByMe, DateTimeOffset? favoritedAt)
    {
        var opened = IsOpened(c.OpenAt);
        string? preview = null;
        if (opened)
        {
            var raw = c.Content.Trim();
            preview = raw.Length > 80 ? raw[..80] + "…" : raw;
        }
        return new CapsuleListItem(
            Id: c.Id.ToString(),
            Code: c.Code,
            Title: c.Title,
            Creator: new UserBrief(owner.Nickname, owner.AvatarId),
            OpenAt: Times.Iso(c.OpenAt),
            CreatedAt: Times.Iso(c.CreatedAt),
            InPlaza: c.InPlaza,
            FavoriteCount: c.FavoriteCount,
            IsOpened: opened,
            FavoritedByMe: favoritedByMe,
            FavoritedAt: favoritedAt is null ? null : Times.Iso(favoritedAt.Value),
            ContentPreview: preview);
    }

    private static bool IsOpened(DateTimeOffset openAt) => openAt <= Times.NowUtc();
}
